### PR TITLE
Create a process instance with start instructions in the Java client

### LINKED ISSUE #9398: Extend the Java Client's CreateProcessInstance API with start instructions

### PR DESCRIPTION
**Pull request by pihme**
## Description

Adds methods to specify start instructions when creating a new process instance.

## Related issues

closes #9398 

## Definition of Done

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [X] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Review by saig0 — COMMENTED** [review]
@pihme thank you for proposing the drafts :+1: 

* second draft: is out because Protobuf classes should not be exposed in the API
* first draft: works and would be okay

However, I would modify your first draft and simplify the API. As a user, I want to write the following commands:

```
// normal instance creation
zeebeClient
   .newCreateInstanceCommand()
   .bpmnProcessId("my-process")
   .latestVersion()
   .variables(json)
   .send();

// create at given element
zeebeClient
   .newCreateInstanceCommand()
   .bpmnProcessId("my-process")
   .latestVersion()
   .variables(json)
   .startBeforeElement("task-1")
   .send();

// create at given elements
zeebeClient
   .newCreateInstanceCommand()
   .bpmnProcessId("my-process")
   .latestVersion()
   .variables(json)
   .startBeforeElement("task-1")
   .startBeforeElement("task-2")
   .send();
```

This API would fit better with our current builder-style API. And it is simpler because you don't need an additional builder class. Regarding multiple instructions, I don't see a need to pass the start instructions as a list.

File: clients/java/src/main/java/io/camunda/zeebe/client/api/command/CreateProcessInstanceCommandStep1.java [hunk 1]
```diff
@@ -101,6 +101,17 @@ interface CreateProcessInstanceCommandStep3 extends FinalCommandStep<ProcessInst
      */
     CreateProcessInstanceCommandStep3 variables(Object variables);
 
+    /**
+     * Overrides the default start position of the process. Calling this method will make the
+     * process start at the given {@code elementId}, if possible. This method can be called more
+     * than once to simultaneously start ar different elements in different branches of the process
```
Comment (saig0) [inline clients/java/src/main/java/io/camunda/zeebe/client/api/command/CreateProcessInstanceCommandStep1.java:107]: :wrench: Fix small typo :writing_hand: 

```suggestion
     * than once to simultaneously start at different elements in different branches of the process.
```

File: clients/java/src/main/java/io/camunda/zeebe/client/api/command/CreateProcessInstanceCommandStep1.java [hunk 2]
```diff
@@ -101,6 +101,17 @@ interface CreateProcessInstanceCommandStep3 extends FinalCommandStep<ProcessInst
...
+     *
+     * @param elementId elementId
```
Comment (saig0) [inline clients/java/src/main/java/io/camunda/zeebe/client/api/command/CreateProcessInstanceCommandStep1.java:109]: :wrench: We could add more description for the parameter.

```suggestion
     * @param elementId the id of the BPMN element where to start the process instance
```

File: clients/java/src/test/java/io/camunda/zeebe/client/impl/command/CreateProcessInstanceCommandImplTest.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+package io.camunda.zeebe.client.impl.command;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.camunda.zeebe.gateway.protocol.GatewayOuterClass.CreateProcessInstanceRequest;
+import io.camunda.zeebe.gateway.protocol.GatewayOuterClass.ProcessInstanceCreationStartInstruction;
+import java.util.List;
+import org.junit.Test;
+
+public class CreateProcessInstanceCommandImplTest {
```
Comment (saig0) [inline clients/java/src/test/java/io/camunda/zeebe/client/impl/command/CreateProcessInstanceCommandImplTest.java:25]: :thinking: I like this unit test in general. But we also have the [CreateProcessInstanceTest](https://github.com/camunda/zeebe/blob/main/clients/java/src/test/java/io/camunda/zeebe/client/process/CreateProcessInstanceTest.java) that contain test cases for the create process instance command. 

Do you want to add a test case to `CreateProcessInstanceTest`?

**Review by saig0 — APPROVED** [review]
@pihme looks good :rocket: 

I have only a few minor suggestions. Please have a look at my comments.

Also, feel free to create a follow-up PR for the additional setter to support multiple instructions for apps like Operate.

**Review by remcowesterhoud — APPROVED** [review]

**Comment by pihme** [comment]
bors merge

**Pull request merged**